dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp

```diff
@@ -0,0 +1,190 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Logger.h>
+#include <Common/TiFlashException.h>
+#include <Flash/Coprocessor/CHBlockChunkCodec.h>
+#include <Flash/Mpp/FineGrainedShuffleWriter.h>
+#include <Flash/Mpp/HashBaseWriterHelper.h>
+#include <Flash/Mpp/MPPTunnelSet.h>
+#include <Flash/Mpp/TrackedMppDataPacket.h>
+
+#include <iostream>
+
+namespace DB
+{
+template <class StreamWriterPtr>
+FineGrainedShuffleWriter<StreamWriterPtr>::FineGrainedShuffleWriter(
+    StreamWriterPtr writer_,
+    std::vector<Int64> partition_col_ids_,
+    TiDB::TiDBCollators collators_,
+    bool should_send_exec_summary_at_last_,
+    DAGContext & dag_context_,
+    uint64_t fine_grained_shuffle_stream_count_,
+    UInt64 fine_grained_shuffle_batch_size_)
+    : DAGResponseWriter(/*records_per_chunk=*/-1, dag_context_)
+    , should_send_exec_summary_at_last(should_send_exec_summary_at_last_)
+    , writer(writer_)
+    , partition_col_ids(std::move(partition_col_ids_))
+    , collators(std::move(collators_))
+    , fine_grained_shuffle_stream_count(fine_grained_shuffle_stream_count_)
+    , fine_grained_shuffle_batch_size(fine_grained_shuffle_batch_size_)
+{
+    rows_in_blocks = 0;
+    partition_num = writer_->getPartitionNum();
+    RUNTIME_CHECK(dag_context.encode_type == tipb::EncodeType::TypeCHBlock);
+    chunk_codec_stream = std::make_unique<CHBlockChunkCodec>()->newCodecStream(dag_context.result_field_types);
+}
+
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::finishWrite()
+{
+    if (should_send_exec_summary_at_last)
+    {
+        batchWriteFineGrainedShuffle<true>();
+    }
+    else
+    {
+        batchWriteFineGrainedShuffle<false>();
+    }
+}
+
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::write(const Block & block)
+{
+    if (block.columns() != dag_context.result_field_types.size())
+        throw TiFlashException("Output column size mismatch with field type size", Errors::Coprocessor::Internal);
+    size_t rows = block.rows();
+    rows_in_blocks += rows;
```

Comment:
Can we extract the code? since I found that is the same with Broadcast writer and HashPartitionWriter

Reply:
no.
From https://github.com/pingcap/tiflash/issues/4693, the batch strategy of different writers should be different